share/completions/meson.fish

```diff
@@ -1,52 +1,308 @@
 # Completions for the meson build system (http://mesonbuild.com/)
 
-set -l basic_arguments \
-    "h,help,show help message and exit" \
-    ",stdsplit,Split stdout and stderr in test logs" \
-    ",errorlogs,Print logs from failing test(s)" \
-    ",werror,Treat warnings as errors" \
-    ",strip,Strip targets on install" \
-    "v,version,Show version number and exit"
-
-set -l dir_arguments \
-    ",localedir,Locale data directory [share/locale]" \
-    ",sbindir,System executable directory [sbin]" \
-    ",infodir,Info page directory [share/info]" \
-    ",prefix,Installation prefix [/usr/local]" \
-    ",mandir,Manual page directory [share/man]" \
-    ",datadir,Data file directory [share]" \
-    ",bindir,Executable directory [bin]" \
-    ",sharedstatedir,Arch-agnostic data directory [com]" \
-    ",libdir,Library directory [system default]" \
-    ",localstatedir,Localstate data directory [var]" \
-    ",libexecdir,Library executable directory [libexec]" \
-    ",includedir,Header file directory [include]" \
-    ",sysconfdir,Sysconf data directory [etc]"
-
-for arg in $basic_arguments
-    set -l parts (string split , -- $arg)
-    if not string match -q "" -- $parts[1]
-        complete -c meson -s "$parts[1]" -l "$parts[2]" -d "$parts[3]"
+function __fish_meson_needs_command
+    set -l cmd (commandline -opc)
+    set -e cmd[1]
+    argparse -s 'v/version' -- $cmd 2>/dev/null
+    or return 0
+    not set -q argv[1]
+end
+
+function __fish_meson_using_command
+    set -l cmd (commandline -opc)
+    set -e cmd[1]
+    test (count $cmd) -eq 0
+    and return 1
+    contains -- $cmd[1] $argv
+    and return 0
+end
+
+function __fish_meson_builddir
+    # Consider the value of -C option to detect the build directory
+    set -l cmd (commandline -opc)
+    argparse -i 'C=' -- $cmd
+    if set -q _flag_C
+        echo $_flag_C
     else
-        complete -c meson -l "$parts[2]" -d "$parts[3]"
+        echo .
     end
 end
 
-for arg in $dir_arguments
-    set -l parts (string split , -- $arg)
-    complete -c meson -l "$parts[2]" -d "$parts[3]" -xa '(__fish_complete_directories)'
+function __fish_meson_targets
+    set -l python (__fish_anypython); or return
+    meson introspect --targets (__fish_meson_builddir) | $python -S -c 'import json, sys
+data = json.load(sys.stdin)
+targets = set()
+for target in data:
+    targets.add(target["name"])
+for name in targets:
+    print(name)' 2>/dev/null
```

Comment:
This doesn't return as many targets as `complete -C"ninja -C $(__fish_meson_builddir) "` does. Are all of the targets discovered by `ninja` usable in this context? If so, maybe we should prefer that. (`ninja` is also faster than meson (python) + python).

(Ninja includes the automatic targets like `benchmark`, `clean`, `install`, `reconfigure`, etc.)

Reply:
`meson compile` doesn't support ninja special targets, just compilation build edges.

Note benchmark is really just a type of test, and it's a bit silly for:
- `meson compile reconfigure` to run ninja and have ninja run `meson setup --reconfigure`.
- `meson compile test` to not compile anything, just run ninja and have ninja run `meson test`
- `meson compile install` to not compile anything, just run ninja and have ninja run `meson install`


clean meanwhile is handled via `meson compile --clean`, it's not a valid target (except directly in ninja).

Reply:
Also, using targets from ninja will not work for backends other than ninja, whereas the current approach does (e.g., it works perfectly for the `xcode` backend).

Reply:
@dseight I wasn't necessarily suggesting using ninja as much as I was wondering about the disparity between the target list. But I did forget about the xcode backend.

@eli-schwartz thanks for clarifying.

Reply:
Yes, that is why `meson compile` exists, essentially. As a portability wrapper for xcode/msbuild compatibility.

If you're happy with ninja, there's no reason to ever use `meson compile`. But it might be nicer in a beginner's README, or you might have xcode-related needs, and that's fine too.

Reply:
> `ninja` is also faster than meson (python) + python

That's true. Calling python twice here isn't a good thing. It actually would be nice to have some convenient builtin to parse json in fish someday, maybe with JSONPath syntax. So this could be reduced to something like:
```fish
meson introspect --targets (__fish_meson_builddir) | json path '$.[name]'
```
This way, we could also replace other python invocations in fish, which are used to parse json.

Reply:
Yeah, we've needed to parse JSON in a number of other completions time and time again. I think if we can find a spiffy and cool way of integrating it into the shell without reimplementing all of `jq` it might be a good addition.